govcd/adminvdc.go

```diff
@@ -52,33 +54,273 @@ func (adminVdc *AdminVdc) Refresh() error {
 // Returns an error if the call to vCD fails.
 // API Documentation: https://vdc-repo.vmware.com/vmwb-repository/dcr-public/7a028e78-bd37-4a6a-8298-9c26c7eeb9aa/09142237-dd46-4dee-8326-e07212fb63a8/doc/doc/operations/PUT-Vdc.html
 func (adminVdc *AdminVdc) UpdateAsync() (Task, error) {
+	apiVersion, err := adminVdc.client.maxSupportedVersion()
+	if err != nil {
+		return Task{}, err
+	}
+	realFunction, ok := vdcProducerByVersion["vdc"+vcdVersionToApiVersion[apiVersion]]
+	if !ok {
+		return Task{}, fmt.Errorf("no entity type found %s", "vdc"+apiVersion)
+	}
+	if realFunction.UpdateVdcAsync == nil {
+		return Task{}, fmt.Errorf("function UpdateVdcAsync is not defined for %s", "vdc"+apiVersion)
+	}
+	util.Logger.Printf("[DEBUG] UpdateAsync call function for version %s", realFunction.SupportedVersion)
 
-	adminVdc.AdminVdc.Xmlns = types.XMLNamespaceVCloud
+	return realFunction.UpdateVdcAsync(adminVdc)
 
-	// Return the task
-	return adminVdc.client.ExecuteTaskRequest(adminVdc.AdminVdc.HREF, http.MethodPut,
-		types.MimeAdminVDC, "error updating VDC: %s", adminVdc.AdminVdc)
 }
 
 // Update function updates an Admin VDC from current VDC struct contents.
 // Any differences that may be legally applied will be updated.
 // Returns an empty AdminVdc struct and error if the call to vCD fails.
 // API Documentation: https://vdc-repo.vmware.com/vmwb-repository/dcr-public/7a028e78-bd37-4a6a-8298-9c26c7eeb9aa/09142237-dd46-4dee-8326-e07212fb63a8/doc/doc/operations/PUT-Vdc.html
 func (adminVdc *AdminVdc) Update() (AdminVdc, error) {
-	task, err := adminVdc.UpdateAsync()
+	apiVersion, err := adminVdc.client.maxSupportedVersion()
```

Comment:
Wondering if we could reuse `UpdateAsync` here instead of function lookup, etc.

Reply:
Internally it's using Async, but here we resolving which function to use.